File: java/src/pattern/observer/rework/StatisticsDisplay.java
package pattern.observer.rework;

import java.util.Observable;
import java.util.Observer;

public class StatisticsDisplay implements Observer, DisplayElement
{
    private float maxTemperature = 0.0f;
    private float minTemperature = 200;
    private float temperatureSum = 0.0f;
    private int numReadings;
    private Observable observable;

    public StatisticsDisplay(Observable observable)
    {
	this.observable = observable;
	observable.addObserver(this);
    }

    public void update(Observable obs, Object arg)
    {
	if (obs instanceof WeatherData) {
	    WeatherData weatherData = (WeatherData)obs;
	    float temperature = weatherData.getTemperature();
	    temperatureSum += temperature;
	    numReadings++;

	    if (temperature > maxTemperature) {
	        maxTemperature = temperature;
	    }

	    if (temperature < minTemperature) {
	        minTemperature = temperature;
	    }

	    display();
	}
    }

    public void display()
    {
	System.out.println("Avg/Max/Min temperature = " + (temperatureSum /
	    numReadings) + " / " + maxTemperature + " / " + minTemperature);
    }
}

File: php/src/pattern/observer/weather/StatisticsDisplay.php
<?php

namespace pattern\observer\weather;

class StatisticsDisplay implements Observer, DisplayElement
{
    protected $_maxTemperature = 0;
    protected $_minTemperature = 200;
    protected $_temperatureSum = 0;
    protected $_numReadings;
    protected $_weatherData;

    public function __construct(Observable $weatherData)
    {
	$this->_weatherData = $weatherData;
	$weatherData->registerObserver($this);
    }

    public function update(Observable $o, $arg = null)
    {
	if ($o instanceof WeatherData) {
	    $this->_temperatureSum += $o->getTemperature();
	    $this->_numReadings++;

	    if ($o->getTemperature() > $this->_maxTemperature) {
		$this->_maxTemperature = $o->getTemperature();
	    }

	    if ($o->getTemperature() < $this->_minTemperature) {
		$this->_minTemperature = $o->getTemperature();
	    }

	    $this->display();
	}
    }

    public function display()
    {
	echo "Avg/Max/Min temperature = "
	    . ($this->_temperatureSum / $this->_numReadings)
	    . " / {$this->_maxTemperature} / {$this->_minTemperature}\n";
    }
}
